src/shogun/machine/gp/SingleFITCLaplacianBase.cpp

```diff
@@ -0,0 +1,409 @@
+		Map<MatrixXd> dKuui(deriv_uu.matrix, deriv_uu.num_rows,
+				deriv_uu.num_cols);
+		Map<MatrixXd> dKui(deriv_tru.matrix, deriv_tru.num_rows,
+				deriv_tru.num_cols);
+
+		ddiagKi*=CMath::sq(m_scale);
+		dKuui*=CMath::sq(m_scale);
+		dKui*=CMath::sq(m_scale);
+
+		result[i]=get_derivative_related_cov(deriv_trtr, deriv_uu, deriv_tru);
+	}
+
+	return result;
+}
+
+float64_t CSingleFITCLaplacianBase::get_derivative_related_mean(SGVector<float64_t> dmu)
+{
+	Map<VectorXd> eigen_al(m_al.vector, m_al.vlen);
+	Map<VectorXd> eigen_dmu(dmu.vector, dmu.vlen);
+	return -eigen_dmu.dot(eigen_al);
+}
+
+SGVector<float64_t> CSingleFITCLaplacianBase::get_derivative_wrt_mean(
+		const TParameter* param)
+{
+	// create eigen representation of al vector
+
+	SGVector<float64_t> result;
+
+	if (param->m_datatype.m_ctype==CT_VECTOR ||
+			param->m_datatype.m_ctype==CT_SGVECTOR)
+	{
+		REQUIRE(param->m_datatype.m_length_y,
+				"Length of the parameter %s should not be NULL\n", param->m_name)
+		result=SGVector<float64_t>(*(param->m_datatype.m_length_y));
+	}
+	else
+	{
+		result=SGVector<float64_t>(1);
+	}
+
+	for (index_t i=0; i<result.vlen; i++)
+	{
+		SGVector<float64_t> dmu;
+
+		dmu=m_mean->get_parameter_derivative(m_features, param, i);
+
+		// compute dnlZ=-dm'*al
+		result[i]=get_derivative_related_mean(dmu);
+	}
+
+	return result;
+}
+
+SGVector<float64_t> CSingleFITCLaplacianBase::get_derivative_wrt_inducing_noise(
+	const TParameter* param)
+{
+	REQUIRE(!strcmp(param->m_name, "inducing_noise"), "Can't compute derivative of "
+			"the nagative log marginal likelihood wrt %s.%s parameter\n",
+			get_name(), param->m_name)
+
+	Map<MatrixXd> eigen_B(m_B.matrix, m_B.num_rows, m_B.num_cols);
+
+	SGMatrix<float64_t> R(m_B.num_rows, m_B.num_cols);
+	Map<MatrixXd> eigen_R(R.matrix, R.num_rows, R.num_cols);
+	//dKuui = 2*snu2; R = -dKuui*B;
+	float64_t factor=2.0*m_ind_noise;
+	eigen_R=-eigen_B*factor;
+
+	SGVector<float64_t> v(m_B.num_cols);
+	Map<VectorXd> eigen_v(v.vector, v.vlen);
+	//v = -sum(R.*B,1)'; 
+	eigen_v=-eigen_R.cwiseProduct(eigen_B).colwise().sum().adjoint();
+
+	SGMatrix<float64_t> dKuui=SGMatrix<float64_t>::create_identity_matrix(m_w.vlen,factor);
+
+	SGVector<float64_t> result(1);
+	//(w'*dKuui*w -al'*(v.*al)- sum(W.*W,1)*v - sum(sum((R*W').*BWt)))/2; 
+	result[0]=get_derivative_related_cov_helper(dKuui, v, R);
+
+	return result;
+}
+
+SGVector<float64_t> CSingleFITCLaplacianBase::get_derivative_related_inducing_features(SGVector<float64_t> v,
+	SGMatrix<float64_t> BdK, const TParameter* param)
+{
+	Map<VectorXd> eigen_v(v.vector, v.vlen);
+	Map<MatrixXd> eigen_B(m_B.matrix, m_B.num_rows, m_B.num_cols);
+	Map<MatrixXd> eigen_BdK(BdK.matrix, BdK.num_rows, BdK.num_cols);
+
+	int32_t dim=((CDotFeatures *)m_inducing_features)->get_dim_feature_space();
+	int32_t num_samples=m_inducing_features->get_num_vectors();
+	SGVector<float64_t>deriv_lat(dim*num_samples);
+	deriv_lat.zero();
+
+	m_lock.lock();
```

Comment:
To avoid this `lock` and gain speed-up, we need three kernel objects via `copy`.  In this case, these three kernel objects will be initialized only once (\Sigma_n, \Sigma_m and \Sigma_{nm}).